Added following features for Product:

Foca22 opened the pull request:
Added following features for Product:
   - add product with category link
   - update product with/without moving to another category
   - delete product
   - get (all) product(s)
   - error handling for not finding product

Renamed all classes which begin with "ProductCategory" to "Category"

Foca22 committed:
Added following features for Product:
   - add product with category link
   - update product with/without moving to another category
   - delete product
   - get (all) product(s)
   - error handling for not finding product

Renamed all classes which begin with "ProductCategory" to "Category"

diana-stoica-ub commented on src/main/java/com/project/store/dto/category/CategoryDto.java:
```diff
@@ -12,6 +12,7 @@ public class ProductCategoryDto {
     @NotEmpty(message = "product.category.type.cannot.be.empty")
     private String type;
```
What should category type be? 
I would also add name, maybe a description

Foca22 replied:
Category type is redundant. 
The initial idea was taken from eMAG where one has: TV, Audio-Video & Foto -> TV & Accesories -> TV
This would overcomplicate matters.
It will be removed

diana-stoica-ub replied:
Name will also be needed.
Regarding category, if you want to make something like this in the future, type will not be enough, you can think of a tree structure, with parent categories.

Foca22 replied:
From the initial commit the name field was accidentally removed. Now is back again in the CategoryDto

diana-stoica-ub commented on src/main/java/com/project/store/service/impl/ProductServiceImpl.java:
```diff
@@ -0,0 +1,78 @@
+        Product updatedProduct = findProductById(productToBeUpdated.getId());
+        updatedProduct.setName(productToBeUpdated.getName());
+        updatedProduct.setPrice(productToBeUpdated.getPrice());
+        updatedProduct.setCategory(productToBeUpdated.getCategory());
```
On the updateProduct action, it would be better only updating the category that the product is assigned to, not the category's properties itself. Please make sure this is how it works.

Foca22 replied:
Removed possibility of updating the name and price. Now only the category can be updated

diana-stoica-ub replied:
Please check again the logic.
What did you want to achieve with this endpoint? What is happening now?

Foca22 replied:
Update action has been modified to only allow the category update of a product together with validations in DTOs

diana-stoica-ub commented on src/main/java/com/project/store/service/impl/ProductServiceImpl.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    public Product updateProductAndCategory(Product productToBeUpdated) {
+        Product updatedProduct = findProductById(productToBeUpdated.getId());
```
Can productId be null? Should it be nullable? If so, what happens in this scenario?

Foca22 replied:
If an invalid productId is sent, then a ProductNotFoundException is thrown with HttpStatus.404
It's a safety mechanism for not sending wrong productIds

diana-stoica-ub replied:
In cases like this, it's better to catch the invalid productId until we query the database, or do any kind of logic with it. If product id si null, an exception should be thrown from the controller level (eg: using spring validators).
If you propagate a null object (that should not be null) downstream, you risk issues like NullPointerExceptions or other unwanted behaviour.

Foca22 replied:
Added @NotNull validation for productId in ProductDto

diana-stoica-ub commented on src/main/java/com/project/store/controller/product/ProductController.java:
```diff
@@ -0,0 +1,70 @@
+
+    @PutMapping("/product-and-category")
+    public ProductDto updateProductAndCategory(@Valid @RequestBody ProductDto productDto) {
+        Category category = categoryService.getCategory(productDto.getCategoryId());
```
Can category id be null? Should it be nullable? If so, what happens in this scenario?

Foca22 replied:
If an invalid categoryId is sent, then a CategoryNotFoundException is thrown with HttpStatus.404
It's a safety mechanism for not sending a wrong categoryIds

diana-stoica-ub replied:
In cases like this, it's better to catch the invalid id until we query the database, or do any kind of logic with it. If category id si null, an exception should be thrown from the controller level (eg: using spring validators).
If you propagate a null object (that should not be null) downstream, you risk issues like NullPointerExceptions or other unwanted behaviour.

Foca22 replied:
Added @NotNull validation to categoryId from CategoryDto

diana-stoica-ub commented on src/main/java/com/project/store/controller/category/CategoryController.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    @GetMapping("/{id}/products")
+    public CategoryWithProductsDto getCategoryWithProducts(@PathVariable Integer id) {
```
I would combine this and /categories-with-products endpoints into just one, pass the categoryId as an optional parameter; if it's set - retrieve only that specific category; if it's null, retrieve all categories.

diana-stoica-ub commented on src/main/java/com/project/store/service/impl/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Override
+    public Category updateCategory(Category categoryToBeUpdated) {
+        Category updatedCategory = findProductCategoryById(categoryToBeUpdated.getId());
```
Can category id be null? Should it be nullable? If so, what happens if it is?

Foca22 replied:
If an invalid categoryId is sent, then a CategoryNotFoundException is thrown with HttpStatus.404
It's a safety mechanism for not sending a wrong categoryIds

diana-stoica-ub replied:
In cases like this, it's better to catch the invalid id until we query the database, or do any kind of logic with it. If category id si null, an exception should be thrown from the controller level (eg: using spring validators).
If you propagate a null object (that should not be null) downstream, you risk issues like NullPointerExceptions or other unwanted behaviour.

Foca22 replied:
Added @NotNull validation to categoryId from CategoryDto

Foca22 committed:
Added following fixes:
-> combining endpoints for gettingCategoriesWithProducts into a single one with an optional parameter for getting:
    - a specific category with its products
    - all categories with all their products
-> removed the possibility of updating the product together with its category to only updating its category

diana-stoica-ub commented on src/main/java/com/project/store/controller/category/CategoryController.java:
```diff
@@ -40,17 +41,15 @@ public List<CategoryDto> getAllCategories() {
-    public List<CategoryWithProductsDto> getAllCategoriesWithProducts() {
-        List<Category> categoryList = categoryService.getAllCategories();
-        return categoryMapper.fromEntitiesToDtosWithProducts(categoryList);
+    @GetMapping(value = {"/products", "/{id}/products"})
```
You can use a non-required @ RequestParam instead.

Foca22 replied:
Added @RequestParam for getting the two results

Foca22 committed:
Added following fixes from the latest review

Foca22 committed:
Small changes

vmircea commented on src/main/java/com/project/store/service/impl/ProductServiceImpl.java:
```diff
@@ -0,0 +1,74 @@
+
+    private CategoryServiceImpl categoryService;
+
+    @Autowired
```
No need to annotate with @autowired if do injection through  constructor

Foca22 replied:
Removed @Autowired annotations from service classes

Foca22 committed:
removed @Autowired annotation from service classes
